library/objective-c/EnvoyEngine.mm

```diff
@@ -23,45 +164,92 @@ + (EnvoyStatus)runWithConfig:(NSString *)config logLevel:(NSString *)logLevel {
   }
 }
 
-+ (EnvoyStream)startStreamWithObserver:(EnvoyObserver *)observer {
-  NSLog(@"%@ not implemented, returning failure", NSStringFromSelector((SEL) __func__));
-  EnvoyStream stream;
-  stream.status = EnvoyStatusFailure;
-  return stream;
++ (void)setupEnvoy {
+  setup_envoy();
+}
+
+// TODO: move to Envoy
++ (EnvoyStream *)startHttpStreamForRequest:(EnvoyRequest *)request handler:(EnvoyObserver *)handler {
+  EnvoyStream *stream = [[EnvoyStream alloc] initWithEngine:engine,
+                                                    request:request,
+                                                   observer:handler];
 }
 
-+ (EnvoyStatus)sendHeaders:(EnvoyHeaders *)headers to:(EnvoyStream *)stream close:(BOOL)close {
-  NSLog(@"%@ not implemented, returning failure", NSStringFromSelector((SEL) __func__));
-  return EnvoyStatusFailure;
+@end
+
+@implementation EnvoyHttpStream
+
+@property (nonatomic, strong) __typeof__(self) strongSelf;
+@property (nonatomic, strong) EnvoyObserver *platformObserver;
+@property (nonatomic, assign) envoy_observer *nativeObserver;
+@property (nonatomic, assign) envoy_stream_t nativeStream;
+
+- (instancetype)initWithObserver:(EnvoyObserver *)observer {
+  self = [super init];
+  if (!self) {
+    return nil;
+  }
+
+  self.platformObserver = observer;
+  envoy_observer *native_obs = (envoy_observer *)malloc(sizeof(envoy_observer));
+
+  atomic_bool *canceled = (atomic_bool *)malloc(sizeof(atomic_bool));
+  atomic_store(canceled, false);
+
+  envoy_observer native_init = {
+    ios_on_headers,
+    ios_on_data,
+    ios_on_trailers,
+    ios_on_metadata,
+    ios_on_complete,
+    ios_on_error,
+    canceled,
+    observer
+  };
+  memcpy(native_obs, &native_init, sizeof(envoy_observer));
+
+  self.nativeObserver = native_obs;
+  envoy_stream result = start_stream(native_obs);
+  if (result.status != ENVOY_SUCCESS) {
+    return nil;
+  }
+
+  self.nativeStream = result.stream;
+  self.strongSelf = self;
+  return self;
 }
 
-+ (EnvoyStatus)sendData:(NSData *)data to:(EnvoyStream *)stream close:(BOOL)close {
-  NSLog(@"%@ not implemented, returning failure", NSStringFromSelector((SEL) __func__));
-  return EnvoyStatusFailure;
+- (void)dealloc {
+  free(self.nativeObserver);
 }
 
-+ (EnvoyStatus)sendMetadata:(EnvoyHeaders *)metadata to:(EnvoyStream *)stream close:(BOOL)close {
-  NSLog(@"%@ not implemented, returning failure", NSStringFromSelector((SEL) __func__));
-  return EnvoyStatusFailure;
+- (void)sendHeaders:(EnvoyHeaders *)headers close:(BOOL)close {
```

Comment:
Header file should be updated to reflect these changes

Reply:
Yep :)